subsys/bluetooth/host/id.c

```diff
@@ -777,6 +777,11 @@ void bt_id_add(struct bt_keys *keys)
 		goto done;
 	}
 
+	/* Delete the identity if it already exists, to make sure that a
+	 * potentially new IRK is stored by hci_id_add.
+	 */
+	bt_id_del(keys);
+
 	err = hci_id_add(keys->id, &keys->addr, keys->irk.val);
```

Comment:
It feels like the deletion of keys is something we should only do if we have actually added them to the controller. 

So it feels like 
A) This should only be done if `CONFIG_BT_PRIVACY` is enabled
and
B) if `keys->state & BT_KEYS_ID_ADDED`

But I haven't enough into this to say more, but it feels weird to unconditionally remove keys. 

Since we are also calling `hci_id_add` (and not `bt_id_add`), should this also have been `hci_id_del` instead?

Reply:
> It feels like the deletion of keys is something we should only do if we have actually added them to the controller.
> 
> So it feels like A) This should only be done if `CONFIG_BT_PRIVACY` is enabled and B) if `keys->state & BT_KEYS_ID_ADDED`
> 
> But I haven't enough into this to say more, but it feels weird to unconditionally remove keys.
> 
> Since we are also calling `hci_id_add` (and not `bt_id_add`), should this also have been `hci_id_del` instead?

Are there cases in which this will go wrong? I thought it only deleted an identity from the resolve list if it existed.

bt_id_del handles a lot of other things that hci_id_del does not, so I though it would be best to use bt_id_del to reverse a bt_id_add that put the existing key there. Not sure though. What do you think?

Reply:
> Since we are also calling `hci_id_add` (and not `bt_id_add`), should this also have been `hci_id_del` instead?

Yes. I'm not sure what `bt_id_del` does exactly. But it clear to me that we want `hci_id_del` here. We want a "`hci_id_add_overwrite`". That is a composition of the two HCI commands `hci_id_del` and `hci_id_add`.

> It feels like the deletion of keys is something we should only do if we have actually added them to the controller.
> 
> So it feels like A) This should only be done if `CONFIG_BT_PRIVACY` is enabled and B) if `keys->state & BT_KEYS_ID_ADDED`
> 
> But I haven't enough into this to say more, but it feels weird to unconditionally remove keys.

I think we want to unconditionally remove the keys here because there is something is wrong with the key management currently. I think of this change as an emergency fix / defense in depth. It does not fix the issue for every case. And it should not be needed if the RL list is properly synced with the existing bonds. I would actually like a `BT_ERR` here when `hci_id_del` returns something other than Unknown Connection Identifier (0x02). It may be a good idea to thoroughly understand why this problem appears in the first case before applying this fix.

Reply:
> Since we are also calling `hci_id_add` (and not `bt_id_add`), should this also have been `hci_id_del` instead?

Yes. I am thinking that we want a  "`hci_id_add_overwrite`". That is a composition of the two HCI commands `hci_id_del` and `hci_id_add`. My reasoning is that the connection between the key management and bond management is slightly broken and there is no event to remove the old keys before adding new keys when a bond replaces another. Assuming the key management overwrites the old key internally (so the sync recovers), it's appropriate to do the same in the controller, so we want `hci_id_add_overwrite`.

On the other hand, if the keys are not correctly updated internally in the bt_id module, something unexpected could happen. We need to test this assumption. Maybe we actually need a `bt_id_add_overwrite` instead.

I think of this change as a stopgap fix. It does not fix the issue for every case. And it should not be needed if the RL list is properly synced with the existing bonds. See https://github.com/zephyrproject-rtos/zephyr/issues/46798#issuecomment-1169908013. I think it's actually appropriate to have a `BT_ERR` here when `hci_id_del` returns something other than Unknown Connection Identifier (0x02). What do you think about that?

Reply:
I think it sounds good to create an hci_id_add_overwrite that deletes the old key if it exists and then adds the new. hci_id_add_overwrite is then called from bt_id_add iinstead of hci_id_add. I will test that approach.

But I do suspect that I run into the same MIC fail issue again because the old LTK is used to decode when re-bonding. If so then I will continue to find out why that is. Please add ideas as to why this might happen.

Reply:
I added an hci_id_overwrite function to id.c that is using hci_id_del and hci_id_add to replace the IRK if it already exists.